component/distributed_cache/distributed_cache.go

```diff
@@ -1277,6 +1281,110 @@ func (dc *DistributedCache) RenameFile(options internal.RenameFileOptions) error
+	log.Debug("DistributedCache::IsDirEmpty: Check if dir is empty: %s", options.Name)
+
+	isAzurePath, isDcachePath, isDebugPath, rawPath := getFS(options.Name)
+	if isDcachePath {
+		log.Debug("DistributedCache::IsDirEmpty: IsDirEmpty for Dcache dir: %s", rawPath)
+		rawPath = filepath.Join(mm.GetMdRoot(), "Objects", rawPath)
+		options.Name = rawPath
+		return dc.NextComponent().IsDirEmpty(options)
+	} else if isAzurePath {
+		log.Debug("DistributedCache::IsDirEmpty: IsDirEmpty for Azure dir: %s", rawPath)
+		options.Name = rawPath
+		return dc.NextComponent().IsDirEmpty(options)
+	} else if isDebugPath {
+		log.Debug("DistributedCache::IsDirEmpty: IsDirEmpty for Debug dir: %s", rawPath)
+		// Debug directories are never empty, they always have some files.
+		return false
+	} else {
+		log.Debug("DistributedCache::IsDirEmpty: IsDirEmpty for Unqualified Path dir: %s", options.Name)
+
+		// We return true if the directory is empty in both dcache and azure.
+		//
+		// Check if directory in empty in dcache.
+		dcachePath := filepath.Join(mm.GetMdRoot(), "Objects", rawPath)
+		options.Name = dcachePath
+		isEmpty := dc.NextComponent().IsDirEmpty(options)
+		if !isEmpty {
+			return false
+		}
+
+		// If the dcache dir is empty, check if the Azure dir is empty.
+		options.Name = rawPath
+		return dc.NextComponent().IsDirEmpty(options)
+	}
+}
+
+// If call comes here, it means that the directory is empty.
+func (dc *DistributedCache) DeleteDir(options internal.DeleteDirOptions) error {
+	log.Debug("DistributedCache::DeleteDir: Delete dir: %s", options.Name)
+
+	isAzurePath, isDcachePath, isDebugPath, rawPath := getFS(options.Name)
+
+	if isDcachePath {
+		log.Debug("DistributedCache::DeleteDir: Delete Dcache dir: %s", rawPath)
+		rawPath = filepath.Join(mm.GetMdRoot(), "Objects", rawPath)
+		options.Name = rawPath
+		err := dc.NextComponent().DeleteDir(options)
+		if err != nil {
+			log.Err("DistributedCache::DeleteDir: Delete failed for Dcache dir %s: %v", options.Name, err)
+			return err
+		}
+	} else if isAzurePath {
+		log.Debug("DistributedCache::DeleteDir: Delete Azure dir: %s", rawPath)
+		options.Name = rawPath
+		err := dc.NextComponent().DeleteDir(options)
+		if err != nil {
+			log.Err("DistributedCache::DeleteDir: Delete failed for Azure dir %s: %v", options.Name, err)
+			return err
+		}
+	} else if isDebugPath {
+		return syscall.EROFS
+	} else {
+		log.Debug("DistributedCache::DeleteDir: Delete Unqualified Path dir: %s", options.Name)
+
+		// Semantics for Unqualified path:
+		// We get this call when both of the directories in (Azure/dcache) are empty.
+		//
+		// Delete Directory from dcache.
+		dcachePath := filepath.Join(mm.GetMdRoot(), "Objects", rawPath)
+		options.Name = dcachePath
+
+		err := dc.NextComponent().DeleteDir(options)
+		if err != nil {
+			log.Err("DistributedCache::DeleteDir: Delete failed for Unqualified Path Dcache dir %s: %v",
+				rawPath, err)
+
+			if err != syscall.ENOENT {
+				return err
+			}
```

Comment:
we should try to delete in both dcache and azure and then in the end return whichever failed